### ISSUE TITLE
How store offline reinforcement learning data (transitions with images , reward etc)

### ISSUE THREAD

**Issue by ugurkanates** [NVIDIA/DALI#2386]
Hello I want to save my data for offline RL/batch RL to load with NVIIDA Dali.
What's best way to do ?

As I know  best way to utilise DALI is with images?
What we are gonna have is that we record images from simulation  and we have other information that is numbers .
What would be best way to record them ? 
I thought about these but i couldnt find any example how to utilise them .

1- Record images as jpeg/png  , record other info .json  or hdf5 or numpy arrayz(npz files) , in reader loop in same batch reading make them (state,action,reward,next_state) lists 
-  Problems  That I have about this :
* Is it possible to make lists from CuPy arrays , because i think in loop thats the variable type of batches ? or are they just tensors(gpu) . If they are just gpu tensors is it still possible to manipulate them like this ?
* How to read additional info with just FileReader ? or NumPy reader ? 
* Does it make sense to two readers in same loop like this ?  But one of our reader would read different images while other just do read one .json file 

2- Record additional info by appending into images as comments or using TIFF file to record them in other layers ? 
And extracting them in reading which would need CPU manipulation right ?  Also if you don't record as images you have to also remove them in order to train with CNN right?

3- Record  (transition sample) as numpy array all together. Still have to extract from whats read no ?

I really have lost about this , i couldnt find any examples. We normallly use TFrecords but we are moving from TFrecords since we are trying to do same thing with PyTorch. So any solution has to be compatible with PyTorch i think

**Comment by JanuszL** [comment]
Hi,
If I understand correctly you want to be able to read images and additional metadata (numerical) using DALI.
You can create a file_list for images and numpy files in the same order and pass them to the corresponding readers, also if you want a random shuffle then you need to set the same seed to both readers.

>2- Record additional info by appending into images as comments or using TIFF file to record them in other layers ?

It may be not the best idea if you are using lossy format, as you would lose the precision of your metadata, but technically it is possible. You can slice the image and slip into image and metadata using the slice operator.

> We normallly use TFrecords but we are moving from TFrecords since we are trying to do same thing with PyTorch.

You can use TFRecord reader with PyTorch and any other FW that DALI supports.

**Comment by ugurkanates** [comment]
> Hi,
> If I understand correctly you want to be able to read images and additional metadata (numerical) using DALI.
> You can create a file_list for images and numpy files in the same order and pass them to the corresponding readers, also if you want a random shuffle then you need to set the same seed to both readers.
> 
> > 2- Record additional info by appending into images as comments or using TIFF file to record them in other layers ?
> 
> It may be not the best idea if you are using lossy format, as you would lose the precision of your metadata, but technically it is possible. You can slice the image and slip into image and metadata using the slice operator.
> 
> > We normallly use TFrecords but we are moving from TFrecords since we are trying to do same thing with PyTorch.
> 
> You can use TFRecord reader with PyTorch and any other FW that DALI supports.

Yeah thanks for the answer. Indeed I want to read image and additional data that is mapped /related to that image together. But not only that during reading I also want to make them transitions ( which is python tuple ) to feed into my network.

What I thought would create problem is that would manipulating data in batch loop of say _next() function have problems ? Since i think data is read directly as CuPy arrays but additional data would be NumPy arrays , so could it be possible to add them to a list together? 

* For TF records I don't know how to save TF records with PyTorch and doesn't seem a good idea to maintain both Tensorflow packages and PyTorch package if we are not using TF for training.

**Comment by JanuszL** [comment]
If you are using `DALIGenericIterator` from `nvidia.dali.plugin.pytorch` then you would have native PyTorch tensors returned, not numpy or cupy arrays. You can check the source code in [this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/use_cases/pytorch/resnet50/pytorch-resnet50.html#) to see how DALI can be integrated with PyTorch based training scrip.
